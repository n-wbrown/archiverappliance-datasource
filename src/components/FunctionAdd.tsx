import React from 'react';
import { reduce, each } from 'lodash';
import { ButtonCascader, CascaderOption } from '@grafana/ui';
import { getCategories } from '../aafunc';
import { FuncDef } from '../types';

export interface FunctionAddProps {
  addFunc: (func: FuncDef) => void;
}

const getAllFunctionNames = (categories: { [key: string]: FuncDef[] }) => {
  const allNames: CascaderOption[] = reduce(
    categories,
    (list, category, key) => {
      const nlist: CascaderOption[] = [];
      each(category, func => nlist.push({ label: func.name, value: func }));
      list.push({ label: key, value: key, children: nlist });
      return list;
    },
    [] as CascaderOption[]
  );

  return allNames;
};

class FunctionAdd extends React.PureComponent<FunctionAddProps> {
  constructor(props: FunctionAddProps) {
    super(props);
  }

  onChange = (value: any[], selectedOptions: CascaderOption[]) => {
    if (value.length < 2) {
      return;
    }
    this.props.addFunc(value[1]);
  };

  render() {
    const categories = getCategories();
    const allFunctions = getAllFunctionNames(categories);
    return (
      <ButtonCascader options={allFunctions} value={undefined} onChange={this.onChange}>
        Add
      </ButtonCascader>
    );
  }
}

export { FunctionAdd };
