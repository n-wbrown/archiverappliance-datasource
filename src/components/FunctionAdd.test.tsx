import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { FunctionAdd, FunctionAddProps } from './FunctionAdd';

const setup = (propOverrides?: object) => {
  const props: FunctionAddProps = {
    addFunc: jest.fn(),
  };

  Object.assign(props, propOverrides);

  return shallow(<FunctionAdd {...props} />);
};

describe('Render', () => {
  it('should render component', () => {
    const wrapper = setup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
