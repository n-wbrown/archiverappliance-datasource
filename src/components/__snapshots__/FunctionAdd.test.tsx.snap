// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render component 1`] = `
<ButtonCascader
  onChange={[Function]}
  options={
    Array [
      Object {
        "children": Array [
          Object {
            "label": "scale",
            "value": Object {
              "category": "Transform",
              "defaultParams": Array [
                "100",
              ],
              "name": "scale",
              "params": Array [
                Object {
                  "name": "factor",
                  "type": "float",
                },
              ],
            },
          },
          Object {
            "label": "offset",
            "value": Object {
              "category": "Transform",
              "defaultParams": Array [
                "100",
              ],
              "name": "offset",
              "params": Array [
                Object {
                  "name": "delta",
                  "type": "float",
                },
              ],
            },
          },
          Object {
            "label": "delta",
            "value": Object {
              "category": "Transform",
              "defaultParams": Array [],
              "name": "delta",
              "params": Array [],
            },
          },
          Object {
            "label": "fluctuation",
            "value": Object {
              "category": "Transform",
              "defaultParams": Array [],
              "name": "fluctuation",
              "params": Array [],
            },
          },
          Object {
            "label": "movingAverage",
            "value": Object {
              "category": "Transform",
              "defaultParams": Array [
                "10",
              ],
              "name": "movingAverage",
              "params": Array [
                Object {
                  "name": "windowSize",
                  "type": "int",
                },
              ],
            },
          },
        ],
        "label": "Transform",
        "value": "Transform",
      },
      Object {
        "children": Array [
          Object {
            "label": "toScalarByAvg",
            "value": Object {
              "category": "Array to Scalar",
              "defaultParams": Array [],
              "name": "toScalarByAvg",
              "params": Array [],
            },
          },
          Object {
            "label": "toScalarByMax",
            "value": Object {
              "category": "Array to Scalar",
              "defaultParams": Array [],
              "name": "toScalarByMax",
              "params": Array [],
            },
          },
          Object {
            "label": "toScalarByMin",
            "value": Object {
              "category": "Array to Scalar",
              "defaultParams": Array [],
              "name": "toScalarByMin",
              "params": Array [],
            },
          },
          Object {
            "label": "toScalarBySum",
            "value": Object {
              "category": "Array to Scalar",
              "defaultParams": Array [],
              "name": "toScalarBySum",
              "params": Array [],
            },
          },
          Object {
            "label": "toScalarByMed",
            "value": Object {
              "category": "Array to Scalar",
              "defaultParams": Array [],
              "name": "toScalarByMed",
              "params": Array [],
            },
          },
          Object {
            "label": "toScalarByStd",
            "value": Object {
              "category": "Array to Scalar",
              "defaultParams": Array [],
              "name": "toScalarByStd",
              "params": Array [],
            },
          },
        ],
        "label": "Array to Scalar",
        "value": "Array to Scalar",
      },
      Object {
        "children": Array [
          Object {
            "label": "top",
            "value": Object {
              "category": "Filter Series",
              "defaultParams": Array [
                "5",
                "avg",
              ],
              "name": "top",
              "params": Array [
                Object {
                  "name": "number",
                  "type": "int",
                },
                Object {
                  "name": "value",
                  "options": Array [
                    "avg",
                    "min",
                    "max",
                    "absoluteMin",
                    "absoluteMax",
                    "sum",
                  ],
                  "type": "string",
                },
              ],
            },
          },
          Object {
            "label": "bottom",
            "value": Object {
              "category": "Filter Series",
              "defaultParams": Array [
                "5",
                "avg",
              ],
              "name": "bottom",
              "params": Array [
                Object {
                  "name": "number",
                  "type": "int",
                },
                Object {
                  "name": "value",
                  "options": Array [
                    "avg",
                    "min",
                    "max",
                    "absoluteMin",
                    "absoluteMax",
                    "sum",
                  ],
                  "type": "string",
                },
              ],
            },
          },
          Object {
            "label": "exclude",
            "value": Object {
              "category": "Filter Series",
              "defaultParams": Array [
                "",
              ],
              "name": "exclude",
              "params": Array [
                Object {
                  "name": "pattern",
                  "type": "string",
                },
              ],
            },
          },
        ],
        "label": "Filter Series",
        "value": "Filter Series",
      },
      Object {
        "children": Array [
          Object {
            "label": "sortByAvg",
            "value": Object {
              "category": "Sort",
              "defaultParams": Array [
                "desc",
              ],
              "name": "sortByAvg",
              "params": Array [
                Object {
                  "name": "order",
                  "options": Array [
                    "desc",
                    "asc",
                  ],
                  "type": "string",
                },
              ],
            },
          },
          Object {
            "label": "sortByMax",
            "value": Object {
              "category": "Sort",
              "defaultParams": Array [
                "desc",
              ],
              "name": "sortByMax",
              "params": Array [
                Object {
                  "name": "order",
                  "options": Array [
                    "desc",
                    "asc",
                  ],
                  "type": "string",
                },
              ],
            },
          },
          Object {
            "label": "sortByMin",
            "value": Object {
              "category": "Sort",
              "defaultParams": Array [
                "desc",
              ],
              "name": "sortByMin",
              "params": Array [
                Object {
                  "name": "order",
                  "options": Array [
                    "desc",
                    "asc",
                  ],
                  "type": "string",
                },
              ],
            },
          },
          Object {
            "label": "sortBySum",
            "value": Object {
              "category": "Sort",
              "defaultParams": Array [
                "desc",
              ],
              "name": "sortBySum",
              "params": Array [
                Object {
                  "name": "order",
                  "options": Array [
                    "desc",
                    "asc",
                  ],
                  "type": "string",
                },
              ],
            },
          },
          Object {
            "label": "sortByAbsMax",
            "value": Object {
              "category": "Sort",
              "defaultParams": Array [
                "desc",
              ],
              "name": "sortByAbsMax",
              "params": Array [
                Object {
                  "name": "order",
                  "options": Array [
                    "desc",
                    "asc",
                  ],
                  "type": "string",
                },
              ],
            },
          },
          Object {
            "label": "sortByAbsMin",
            "value": Object {
              "category": "Sort",
              "defaultParams": Array [
                "desc",
              ],
              "name": "sortByAbsMin",
              "params": Array [
                Object {
                  "name": "order",
                  "options": Array [
                    "desc",
                    "asc",
                  ],
                  "type": "string",
                },
              ],
            },
          },
        ],
        "label": "Sort",
        "value": "Sort",
      },
      Object {
        "children": Array [
          Object {
            "label": "maxNumPVs",
            "value": Object {
              "category": "Options",
              "defaultParams": Array [
                "100",
              ],
              "name": "maxNumPVs",
              "params": Array [
                Object {
                  "name": "number",
                  "type": "int",
                },
              ],
            },
          },
          Object {
            "label": "binInterval",
            "value": Object {
              "category": "Options",
              "defaultParams": Array [
                "900",
              ],
              "name": "binInterval",
              "params": Array [
                Object {
                  "name": "interval",
                  "type": "int",
                },
              ],
            },
          },
          Object {
            "label": "disableAutoRaw",
            "value": Object {
              "category": "Options",
              "defaultParams": Array [
                "true",
              ],
              "name": "disableAutoRaw",
              "params": Array [
                Object {
                  "name": "boolean",
                  "options": Array [
                    "true",
                    "false",
                  ],
                  "type": "string",
                },
              ],
            },
          },
          Object {
            "label": "disableExtrapol",
            "value": Object {
              "category": "Options",
              "defaultParams": Array [
                "true",
              ],
              "name": "disableExtrapol",
              "params": Array [
                Object {
                  "name": "boolean",
                  "options": Array [
                    "true",
                    "false",
                  ],
                  "type": "string",
                },
              ],
            },
          },
        ],
        "label": "Options",
        "value": "Options",
      },
    ]
  }
>
  Add
</ButtonCascader>
`;
